Added MIDI Remap Lua DSP plugin

Linux-cpp-lisp opened the pull request:
Hi all,

This is an Ardour Lua DSP script I put together for remapping arbitrary MIDI notes in real time. I [posted it on the forums](https://discourse.ardour.org/t/lua-dsp-plugin-for-real-time-midi-remapping/100475), and at the suggestion of Robin Gareus, I'm submitting a PR to have it added to Ardour.

It can map any MIDI note to any other. Polyphonic aftertouch should also be remapped, though I haven't tested that, lacking any easy way to produce polyphonic aftertouch data. Timing and velocities are unaffected, and all other MIDI data is passed through untouched. By default it allows up to 10 remappings; just change `N_REMAPINGS` in the code if you need more.

One major improvement -- if possible -- would be to update the `translation_table` only when the parameters have changed, rather than every single buffer. I was unable to find any callback for when the parameters have been modified, but if one does exist, I'll certainly update the code to take advantage of it. 

Thanks,
A.

x42 commented on scripts/midi_remap.lua:
```diff
@@ -0,0 +1,103 @@
+    -- We build the translation table every buffer because, as far as I can tell,
+    -- there's no way to only rebuild it when the parameters have changed.
+    -- As a result, it has to be updated every buffer for the parameters to have
+    -- any effect.
```
You could store it globally and only update it when a parameter has changed. Not sure if it's worth it though.

An alternative is to not initialize the table with a unity matrix but use Lua's sparse array. 

```
  local translation_table = {}
  local ctrl = CtrlPorts:array()
  for i=1,N_REMAPINGS*2,2 do
    translation_table[ctrl[i]] = ctrl[i + 1]
  end
-- .....
  d[2] = translation_table[d[2]] or d[2]
```

Linux-cpp-lisp replied:
Is there a callback for when the parameters have changed? If not, I would have to store the past parameter values and compare them, which takes nearly as many operations as rebuilding the table and introduces an extra branch, so I doubt it would actually change performance (I know that's C-level nitpicking, but either way I doubt the differences are measurable at this level).

I opted against the sparse array since my research indicated that tables indexed with contiguous positive integers were implemented under the hood as C-style arrays, and as a result would hopefully be efficient... but since it looks like most Lua versions [do a hash table anyway](https://stackoverflow.com/questions/26843589/how-does-lua-5-x-represent-sparse-arrays), maybe it's better to just clean up the code and use one.

Linux-cpp-lisp replied:
Whoops. Actually, the sparse array won't help: say there is a mapping from 2 -> 3 that is later changed to 5 -> 3, for example. If `translation_table` isn't reinitialized to the identity, now *both* 2 and 5 will map to 3... I think I'll leave it as is.

(I'm also pretty sure it's already a global -- omitting `local` in the declaration is sufficient, right?)

x42 replied:
> Is there a callback for when the parameters have changed? If not, I would have to store the past parameter values and compare them

yes, you would have to store and compare. It's not worth it.

>  If translation_table isn't reinitialized to the identity, ..

`local translation_table = {}` initializes it -- you can also get rid of the the `dsp_init` function and global definition.

x42 commented on scripts/midi_remap.lua:
```diff
@@ -0,0 +1,103 @@
+    -- Update it
+    local ctrl = CtrlPorts:array()
+    for i=1,N_REMAPINGS*2,2 do
+        translation_table[ctrl[i]] = ctrl[i + 1]
```
If there are duplicate mappings, the last one wins.

This is mainly relevant for the default  `000 -> n`.  Try re-mapping note 0 using any but the last control.
One solution would be to add a dedicated  `"off" == -1` case as control parameter.

Linux-cpp-lisp replied:
Since I'm switching to the sparse arrays, I'll add an dedicated "off" option

Linux-cpp-lisp commented on scripts/midi_remap.lua:
```diff
@@ -0,0 +1,103 @@
+        local octave = -5 + math.floor(note / 12)
+        local name = note_names[(note % 12) + 1]
+        map_scalepoints[string.format("%03d (%s%d)", note, name, octave)] = note
+    end
```
Great! I didn't know that function existed, but it's much better than this ad-hoc approach.

Linux-cpp-lisp commented:
OK, I've resolved the edge cases, switched to the sparse `translation_table`, added an "off" option, and switched to the localized note names. Thanks for the feedback @x42 !

x42 commented:
Now "to" can be "off" as well.  I'd expect that to drop the note.

What do you think? should "to" not be able to be "off", or should a new feature be added to filter, drop notes in that case?

Linux-cpp-lisp commented:
Good idea -- I've changed it so that mapping a note to "Off" (which I've renamed "None" for clarity) simply filters out the events, ` > /dev/null` style.

Great!

x42 commented:
Awesome! rebased and merged as 6.0-pre0-1578 f0d227f92

Thanks!
